lib/native/mqtt.ts

```diff
@@ -237,29 +245,20 @@ export class MqttClientConnection extends NativeResourceMixin(BufferedEventEmitt
         return new Promise<boolean>((resolve, reject) => {
             reject = this._reject(reject);
 
-            const on_connect = (error_code: number, return_code: number, session_present: boolean) => {
-                if (error_code == 0 && return_code == 0) {
-                    resolve(session_present);
-                    this.emit('connect', session_present);
-                } else if (error_code != 0) {
-                    reject("Failed to connect: " + io.error_code_to_string(error_code));
-                } else {
-                    reject("Server rejected connection.");
-                }
-            }
-
             try {
                 crt_native.mqtt_client_connection_connect(
                     this.native_handle(),
                     this.config.client_id,
                     this.config.host_name,
                     this.config.port,
+                    resolve,
+                    reject,
```

Comment:
After thinking about this for 2 days, what if we used [bind](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_objects/Function/bind) to create the single closure that needs to be passed to native code here. I really, really don't like the extra ref counting in native code, especially when all of the references have the same lifetime.
e.g.
```typescript
this._on_connect_callback.bind(this, resolve, reject)
```

Reply:
I'll give it a try and see how it goes.